### PR TITLE
Create post-install script for IBM Cloud machines

### PR DESCRIPTION
**Pull request by mkowalski**
This PR creates a post-install script that can be used on the IBM Cloud
machines so that they are deployed with a set of reasonable defaults.

### CODE REVIEW

**Comment by mkowalski** [comment]
A thing that may be spotted but is not really caused by this PR is the fact that the dnsmasq instance controlled by the NetworkManager does not start on `systemctl restart NetworkManager.service` nor on reboot, but only on `systemctl reload NetworkManager.service`.

For this reason, after rebooting the server a manual `systemctl reload NetworkManager.service` is needed, otherwise there is no `nameserver 127.0.0.1` entry in `/etc/resolv.conf` and there is no `dnsmasq` process running on the host.

Given that it's kinda out of scope of the provisioning script, I'm keeping it as it is until I find a proper place to plug a workaround for this.

Ref.: https://serverfault.com/questions/1054792/networkmanager-isnt-starting-dnsmasq-on-restart-only-reload
Ref.: https://www.reddit.com/r/CentOS/comments/lrv20x/networkmanager_isnt_starting_dnsmasq_on_restart/

File: scripts/ibmcloud_post_install.sh [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+log "Starting ibmcloud-post-install.sh"
+
+### SSH hardening
+### IBM Cloud machines by default allow SSH as root using the password. Given that the default one
+### is extremely simple, we disable this option.
+
+sed -i "s/.*RSAAuthentication.*/RSAAuthentication yes/g" /etc/ssh/sshd_config
+sed -i "s/.*PubkeyAuthentication.*/PubkeyAuthentication yes/g" /etc/ssh/sshd_config
+sed -i "s/.*PasswordAuthentication.*/PasswordAuthentication no/g" /etc/ssh/sshd_config
```
Comment (celebdor) [inline scripts/ibmcloud_post_install.sh:19]: Could pass the multiple replacements to a single sed command.

File: scripts/ibmcloud_post_install.sh [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+systemctl restart sshd.service
+
+log "SSH configuration applied successfuly"
+
+### SSH local key
+
+mkdir -p /root/.ssh
+chmod 700 /root/.ssh
+ssh-keygen -t rsa -b 4096 -N "" -f /root/.ssh/id_rsa
+cat /root/.ssh/id_rsa.pub >> /root/.ssh/authorized_keys
```
Comment (celebdor) [inline scripts/ibmcloud_post_install.sh:30]: What do we need to authorize the local root key for?

File: scripts/ibmcloud_post_install.sh [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+dnf install -y dnsmasq
+
+mkdir -p /etc/NetworkManager/conf.d/
+cat <<EOF >/etc/NetworkManager/conf.d/dnsmasq.conf
+[main]
+dns = dnsmasq
+EOF
+
+cat /etc/resolv.conf  | grep -v '^#' | grep nameserver | awk '{print "server="$2}' > /etc/NetworkManager/dnsmasq.d/ibm-cloud.conf
```
Comment (celebdor) [inline scripts/ibmcloud_post_install.sh:58]: Why shouldn't it just be:

```
awk '/^nameserver / {print "server="$2}' /etc/resolv.conf >> /etc/NetworkManager/dnsmasq.d/ibm-cloud.conf

File: scripts/ibmcloud_post_install.sh (hunk 2 above)
Comment (mkowalski) [reply scripts/ibmcloud_post_install.sh:30]: I often find myself doing `ssh localhost`

File: scripts/ibmcloud_post_install.sh (hunk 3 above)
Comment (mkowalski) [reply scripts/ibmcloud_post_install.sh:58]: Yes it could and should. Fixed :)

File: scripts/ibmcloud_post_install.sh (hunk 1 above)
Comment (mkowalski) [reply scripts/ibmcloud_post_install.sh:19]: Purely for readability, using multiple expressions in one will create a big one-liner difficult to modify in the future if someone wants to add one more parameter to change. With this here, you can just copy-paste and change a single one so `git blame` and stuff like this becomes easier

**Comment by YuviGold** [comment]
Could you elaborate on what's the difference between this script and the one running with `make create_full_environment`? This is the script we use for Packet instances, if it's not generic enough let's adjust it

**Comment by mkowalski** [comment]
> Could you elaborate on what's the difference between this script and the one running with `make create_full_environment`? This is the script we use for Packet instances, if it's not generic enough let's adjust it

The whole of `./create_full_environment.sh` is too big because provisioning script can be one script, it cannot be a bunch of files sourcing each other. If we take `./install_environment.sh` instead it seems to be like it's a bit too heavy because it's designed specifically for assisted-test-infra and does stuff that I do not need (or even, must not do) if I'm using the server as a workstation or for dev-infra or for any other work (e.g. `setenforce 0` is something I won't do unless really needed).

**Comment by romfreiman** [comment]
@mkowalski as I see it, the purpose of those servers are for assisted development, where 100% of the people use assisted-test-infra.
I dont think we should maintain 2 scripts. Most of the people dont care about setenforce 0. So this is not an issue for them. If it's really important for you, feel free top change this post deployement. IMO it's still less maintenance than having 2 scripts

**Comment by mkowalski** [comment]
Sure, but this is not the main point of the discussion and I don't want to diverge into parts which are implementation details (yet). The problem here is not *what* the script is going to do but *how* to run it. In the current form you *cannot* use `./create_full_environment.sh`.

If we use only `./install_environment.sh` as a provisioning script, then you would still need to run `make create_full_environment` afterwards because e.g. minikube has not been installed.

**Comment by mkowalski** [comment]
@romfreiman @YuviGold I have reorganized the structure base on what we have discussed

/hold "it needs a test"

**Comment by mkowalski** [comment]
The first non-interactive run gave me

```
  Installing       : seabios-bin-1.13.0-2.module_el8.4.0+547+a85d02ba    85/123 
  Running scriptlet: radvd-2.17-15.el8.x86_64                            86/123 
error: %prein(radvd-2.17-15.el8.x86_64) scriptlet failed, signal 15

  Installing       : mesa-filesystem-20.3.3-2.el8.x86_64                 87/123 
error: radvd-2.17-15.el8.x86_64: install failed
```

File: scripts/ibmcloud_post_install.sh [hunk 4]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env bash
+
+cat > /root/install_complete.sh <<EOF
```
Comment (tsorya) [inline scripts/ibmcloud_post_install.sh:3]: why not /usr/bin?

File: scripts/ibmcloud_post_install.sh [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+log "Starting ibmcloud-post-install.sh"
+
+{
+### Base prerequisites
+dnf install -y git
+
+### Get AI repo
+cd /tmp
```
Comment (tsorya) [inline scripts/ibmcloud_post_install.sh:22]: maybe better to clone it to /home ?

Comment (tsorya) [reply scripts/ibmcloud_post_install.sh:22]: i mean why to remove it? if i understand right those machines will be used for test-infra in any case right?

File: scripts/ibmcloud_post_install.sh (hunk 4 above)
Comment (mkowalski) [reply scripts/ibmcloud_post_install.sh]: Done

**Pull request merged**